File: 1717_집합의 표현.cpp
#include <iostream>
#include <algorithm>
#include <stack>
#include <queue>
#include <vector>
#include <set>
#include <map>
#include <cmath>
#include <cstring>
#define pii pair<int,int>
#define INF 1e9+7
#define MOD 1004
typedef long long ll;
using namespace std;

void init();
int find(int);
bool UNION(int,int);
 
int n,m;
int parent[1000001];
int height[1000001];

int main(){
    ios_base::sync_with_stdio(false); 
    cin.tie(0); cout.tie(0);

    init();
    for(int i=0;i<m;i++){
        int a,b,c;
        cin>>a>>b>>c;
        if(a==0) UNION(b,c);
        else if(find(b)==find(c))cout<<"YES\n";
        else cout<<"NO\n";
    }
    
    return 0;
}

void init(){
    cin>>n>>m;
    memset(&parent,-1,sizeof(parent));
}

int find(int x){
    if(parent[x]==-1) return x;
    return parent[x]=find(parent[x]);
}

bool UNION(int x, int y){
    x=find(x);
    y=find(y);
    if(x==y) return false;
    if(height[x]<height[y]) swap(x,y);
    parent[y]=x;
    if(height[x]==height[y]) height[y]++;
    return true;
}

File: 1717_집합의표현.py
import sys

def find(x) :
    if x == parent[x] : return x
    parent[x] = find(parent[x])
    return parent[x]

def union(x, y) :
    nx = find(x)
    ny = find(y)
    if nx == ny :
        return
    if height[nx] < height[ny] :
        nx, ny = ny, nx
    parent[ny] = nx
    if height[nx] == height[ny] :
        height[ny] += 1

n, m = map(int, sys.stdin.readline().split())
parent = [i for i in range(n + 1)]
height = [0] * (n+1)
for _ in range(m) :
    mode, a, b = map(int, sys.stdin.readline().split())
    if mode == 0 :
        union(a,b)
    elif find(a) == find(b) :
        print("YES")
    else :
        print("NO")
    